cpp/lib/ResultSetArrow.cpp

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2021 Snowflake Computing, Inc. All rights reserved.
+ */
+
+#include <cmath>
+#include <iomanip>
+#include <memory>
+#include <sstream>
+#include <time.h>
+
+#include "../logger/SFLogger.hpp"
+#include "ResultSetArrow.hpp"
+#include "result_set_arrow.h"
+#include "results.h"
+
+
+namespace Snowflake
+{
+namespace Client
+{
+
+
+ResultSetArrow::ResultSetArrow() :
+    Snowflake::Client::ResultSet()
+{
+    m_queryResultFormat = QueryResultFormat::ARROW;
+    m_bufferBuilder = std::make_shared<arrow::BufferBuilder>();
+}
+
+ResultSetArrow::ResultSetArrow(
+    cJSON * initialChunk,
+    SF_COLUMN_DESC * metadata,
+    std::string tzString
+) :
+    ResultSet(metadata, tzString)
+{
+    m_queryResultFormat = QueryResultFormat::ARROW;
+    m_bufferBuilder = std::make_shared<arrow::BufferBuilder>();
+    init(initialChunk);
+}
+
+// Public methods ==================================================================================
+
+SF_STATUS STDCALL ResultSetArrow::appendChunk(cJSON * chunk)
+{
+    // Decode Base64-encoded Arrow-format rowset of the chunk.
+    const char * base64Rowset = snowflake_cJSON_GetStringValue(chunk);
```

Comment:
The JSON response from server has a key value pairs,  and rowsetBase64 is one of the Key and its value is the base64 encoded data. 
The function snowflake_cJSON_GetStringValue does not return the base64 encoded data.

Reply:
```
{
  "data" : {
    "parameters" : [{"name":"CLIENT_PREFETCH_THREADS","value":4},{"name":"TIMESTAMP_OUTPUT_FORMAT","value":"DY, DD MON YYYY HH24:MI:SS TZHTZM"},{"name":"TIME_OUTPUT_FORMAT","value":"HH24:MI:SS"},{"name":"CLIENT_RESULT_CHUNK_SIZE","value":160},{"name":"TIMESTAMP_TZ_OUTPUT_FORMAT","value":""},{"name":"ODBC_SCHEMA_CACHING","value":false},{"name":"CLIENT_SESSION_KEEP_ALIVE","value":false},{"name":"ODBC_TREAT_BIG_NUMBER_AS_STRING","value":false},{"name":"CLIENT_OUT_OF_BAND_TELEMETRY_ENABLED","value":false},{"name":"CLIENT_METADATA_USE_SESSION_DATABASE","value":false},{"name":"TIMESTAMP_NTZ_OUTPUT_FORMAT","value":""},{"name":"CLIENT_RESULT_PREFETCH_THREADS","value":1},{"name":"CLIENT_METADATA_REQUEST_USE_CONNECTION_CTX","value":false},{"name":"CLIENT_HONOR_CLIENT_TZ_FOR_TIMESTAMP_NTZ","value":true},{"name":"CLIENT_MEMORY_LIMIT","value":1536},{"name":"ODBC_ENABLE_ARROW_QUERY_RESULT_FORMAT","value":true},{"name":"ODBC_ENABLE_COMPRESSION","value":true},{"name":"CLIENT_TIMESTAMP_TYPE_MAPPING","value":"TIMESTAMP_LTZ"},{"name":"TIMEZONE","value":"UTC"},{"name":"ODBC_TREAT_DECIMAL_AS_INT","value":false},{"name":"SERVICE_NAME","value":""},{"name":"CLIENT_RESULT_PREFETCH_SLOTS","value":2},{"name":"ODBC_USE_BULK_FETCHING","value":true},{"name":"ODBC_PUSHDOWN_FILTERS","value":true},{"name":"ODBC_ENABLE_PUT_GET","value":true},{"name":"CLIENT_TELEMETRY_ENABLED","value":true},{"name":"CLIENT_USE_V1_QUERY_API","value":true},{"name":"CLIENT_DISABLE_INCIDENTS","value":true},{"name":"CLIENT_RESULT_COLUMN_CASE_INSENSITIVE","value":false},{"name":"BINARY_OUTPUT_FORMAT","value":"HEX"},{"name":"CLIENT_ENABLE_LOG_INFO_STATEMENT_PARAMETERS","value":false},{"name":"CLIENT_FORCE_PROTECT_ID_TOKEN","value":true},{"name":"DATE_OUTPUT_FORMAT","value":"YYYY-MM-DD"},{"name":"CLIENT_CONSENT_CACHE_ID_TOKEN","value":false},{"name":"CLIENT_STAGE_ARRAY_BINDING_THRESHOLD","value":65280},{"name":"CLIENT_SESSION_KEEP_ALIVE_HEARTBEAT_FREQUENCY","value":3600},{"name":"CLIENT_SESSION_CLONE","value":false},{"name":"AUTOCOMMIT","value":false},{"name":"TIMESTAMP_LTZ_OUTPUT_FORMAT","value":""},{"name":"ODBC_ENABLE_COMBINED_DESCRIBE","value":true},{"name":"ODBC_USE_CUSTOM_SQL_DATA_TYPES","value":false},{"name":"ODBC_USE_NEW_JSON_PARSER2","value":true} ],
    "rowtype" : [{"name":"LEFT('LEFT IS RETURNED', 16)","database":"","schema":"","table":"","scale":null,"precision":null,"length":16,"type":"text","nullable":false,"byteLength":64,"collation":null} ],
    
"rowsetBase64" : "/////1ABAAAQAAAAAAAKAA4ABgANAAgACgAAAAAABAAQAAAAAAEKAAwAAAAIAAQACgAAAAgAAAAIAAAAAAAAAAEAAAAYAAAAAAASABgAFAATABIADAAAAAgABAASAAAAFAAAAMQAAADIAAAAAAAFAcQAAAAEAAAAiAAAAFgAAAAsAAAABAAAAJD///8IAAAADAAAAAIAAAAxNgAACgAAAGNoYXJMZW5ndGgAALT///8IAAAAEAAAAAQAAABURVhUAAAAAAsAAABsb2dpY2FsVHlwZQDc////CAAAAAwAAAACAAAANjQAAAoAAABieXRlTGVuZ3RoAAAIAAwACAAEAAgAAAAIAAAADAAAAAMAAABMT0IADAAAAHBoeXNpY2FsVHlwZQAAAAAAAAAABAAEAAQAAAAcAAAATEVGVCgnTEVGVCBJUyBSRVRVUk5FRCcsIDE2KQAAAAD/////wAAAAAAAAAAAAAAEAAAACAAAAAwAAAAQAAAAFAAAABgAAAAcAAAAIAAAACQAAAAoAAAALAAAADAAAAA0AAAAOAAAADwAAAAAAEAABABAAAgAQAAMAEAAEABAABQAQAAYAEAAHABAACAAQAAkAEAAKABAACwAQAAwAEAANABAADgAQAA8AEAAAACAAAQAgAAIAIAADACAABAAgAAUAIAAGACAABwAgAAgAIAAJACAACgAgAAsAIAAMACAADQAgAA4AIAAPACAAAAAwAAEAMAACADAAAwAwAAQAMAAFADAABgAwAAcAMAAIADAACQAwAAoAMAALADAADAAwAA0AMAAOADAADwAwAAAAQAABAEAAAgBAAAMAQAAEAEAABQBAAAYAQAAHAEAACABAAAkAQAAKAEAACwBAAAwAQAANAEAADgBAAA8AQAAAAFAAAQBQAAIAUAADAFAABABQAAUAUAAGAFAABwBQAAgAUAAJAFAACgBQAAsAUAAMAFAADQBQAA4AUAAPAFAAAABgAAEAYAACAGAAAwBgAAQAYAAFAGAABgBgAAcAYAAIAGAACQBgAAoAYAALAGAADABgAA0AYAAOAGAADwBgAAAAcAABAHAAAgBwAAMAcAAEAHAABQBwAAYAcAAHAHAACABwAA”,
    "total" : 5000,
    "returned" : 5000,
    "queryId" : "01999ca4-0391-126f-0000-000800016e32",
    "databaseProvider" : null,
    "finalDatabaseName" : "TESTDB",
    "finalSchemaName" : "SFTEST_001",
    "finalWarehouseName" : "REGRESS",
    "finalRoleName" : "SYSADMIN",
    "numberOfBinds" : 0,
    "arrayBindSupported" : false,
    "statementTypeId" : 4096,
    "version" : 1,
    "sendResultTime" : 1610659476659,
    "queryResultFormat" : "arrow"
  },
  "code" : null,
  "message" : null,
  "success" : true
}
```
Server response looks lime this. And the base64Rowset is a key and its value has the encoded data.

Reply:
Sorry, not sure how that got through.
This has been modified to use `snowflake_cJSON_GetObjectItem()` to retrieve the base64 rowset.